### ISSUE TITLE
Feature: Store parsed html document in visit object

### ISSUE THREAD

**Issue by ScoobyDid** [swup/swup#862]
### Describe the problem 🧐

The swup.js library currently parses the entire HTML document internally using the DOMParser API. However, the parsed HTML document is not exposed or accessible to the users of the library. The library only returns the HTML document as a string.

Parsing an HTML document is a relatively slow process. In scenarios where the developer needs to work with the parsed HTML document, they have to parse the same HTML document again. This results in unnecessary duplication of work and performance overhead.

To improve the efficiency and flexibility of the library, I propose adding an option that allows users to control whether the parsed HTML document should be included in the visit object or not. When this option is enabled, the visit object should include a reference to the parsed HTML document.

I believe this feature would be a valuable addition to the swup.js library and would greatly improve its usability and performance in certain scenarios.

Thank you for considering this feature request.

### Describe the propsed solution 😎

const swup = new Swup({
  includeParsedHtmlInVisit: true, // New option
  // Other options...
});

swup.hooks.on('content:replace', (visit) => {
  const parsedHtml = visit.to.html.parsed; // Parsed HTML document
  // ...
});

### Alternatives considered 🤔

- Parse the same document string again
- Don't parse it again - search in string using regex

### How important is this feature to you? 🧭

Nice to have

### Checked all these? 📚

- [X] I have read and searched the [official docs](https://swup.js.org)
- [X] I have checked [discussions](https://github.com/orgs/swup/discussions) and [existing issues](https://github.com/issues?q=is:issue+org:swup) for related problems
- [X] I agree to follow this project's [Code of Conduct](https://github.com/swup/.github/blob/master/CODE_OF_CONDUCT.md)
- [X] I have provided all necessary information to the best of my knowledge

**Comment by daun** [comment]
I like the idea! I've actually been toying with the same thought for a while. It'd certainly be great for performance to only parse the document once. There's two things to consider here:

1. Storing the full document in the visit object will probably lead to inadvertent memory leaks. Event handlers holding onto the visit object will now also prevent all those parsed documents from being garbage collected. We've made an effort to only store simple data structures like strings and arrays in the visit object for that reason.
2. Some plugins and event handlers might modify the response html on the fly. Having an html string and an html document would require making any changes in both places, leading to subtle bugs.

I've previously felt that these are reason enough to avoid this for now, but there might be great solutions for either of them? Open for suggestions!

**Comment by ScoobyDid** [comment]
I understand the reason now. leaking such thing may result in even worse performance than parsing document another time and then letting garbage collector do the work; i have thought about possible optimizations - here are some of them:

1. If implemented, this option of course should be off by default - to be enabled only when needed
2. We can remove swup element(s) from parsed document - they are anyway placed in document; their html is accessible in document after "content:replace" hook so there is no need to also store them in the visit object. In most cases this will greatly reduce size of document that is saved in visit object, because, for example, there often is just one <main id="swup"> element, so, if we will save all but this element, there will be just <head> and maybe  <header> and <footer> saved in memory. Code may look something like this:
```
const incomingDocument = new DOMParser().parseFromString(html, 'text/html');
// ... after some stuff done:
incomingDocument.querySelectorAll(this.options.containers.join(',')).forEach(el => el.remove())
```
3. after we done with parsed html document, either from library or from library user side incomingDocument  should be set to null; this will increase chances that it will be garbage collected

Although these heuristics will not completely solve the problem, especially regarding possible bugs that it can cause; fixing this is more challenging than i thought - as always :) . Thanks anyway, swup is really nice tiny & flexible enough library to be used almost on any website

**Comment by daun** [comment]
Appreciate the input. Regarding point 3 — we could probably get away with setting `visit.to.document = null` after finishing or cancelling a visit. Any memory leaks will then definitely be caused by the consumer and would have occurred anyway.

Before we go to deep on this — have you done any measurements on the performance of DomParser by chance? Sometimes, we get ahead of ourselves trying to optimize things that don't need optimizing. If parsing the document only takes a millisecond or two, I can see no harm in doing it two or even three times.

**Comment by ScoobyDid** [comment]
I have done some simple tests, on multiple devices, with the help of browserstack; all tests have been done in clean environment, without extensions; i have tested it on 2 pages: 1 - current github page; 2 - [page with simple html structure](https://wave.webaim.org/). i have tested using following code:

```
function measure(amount) {
  const start = performance.now();
  for (let i = 0; i < amount; i++) {
    const parsed = new DOMParser().parseFromString(html, 'text/html');
  }
  const end = performance.now() - start;
  console.log(`${amount}: ${end}`);
  document.body.insertAdjacentHTML('afterbegin', `<div>${amount}: ${end}</div>`);
}

measure(1);
measure(3);
measure(10);
```

As you see, i have measured parsing document 1, 3 and 10 times; I have tested it 3 times on each device. Here are the results in milliseconds:


[current page:](https://github.com/swup/swup/issues/862:)

mid-end notebook (intel core i3 1005g1), chrome 120:
1:   55; 41; 37;
3:   124; 92; 90;
10: 388; 363; 285;

galaxy g5 (low-end mobile), chrome 119:
1:  369; 286; 307;
3:  1131; 815; 808;
10: 3102; 2713; 2709;

galaxy s9 (mid-end mobile), chrome 120:
1:  104; 101; 103;
3:  270; 264; 257; 
10: 758; 775; 772;

galaxy s22 (high-end mobile), chrome 120:
1:  80; 71; 73;
3:  194; 200; 195;
10: 630; 688; 689;

pixel 6, chrome 120:
1:  49; 43; 47;
3:  128; 130; 130;
10: 426; 430 431;

iphone 13, safari 17:
1:  20; 23; 23;
3:  41; 43; 42;
10: 130; 130; 131;

iphone 13, chrome 120:
1:  18; 18; 19;
3:  39; 38; 39;
10: 123; 122; 123;



[other page:](https://wave.webaim.org/)


mid-end notebook (intel core i3 1005g1), chrome 120:
1: 22; 32; 16;
3: 58; 45; 36;
10: 176; 118; 126;

galaxy g5 (low-end mobile), chrome 119:
1: 342; 156; 160;
3: 857; 437; 422;
10: 2179; 1411; 1416;

galaxy s9 (mid-end mobile), chrome 120:
1: 65; 60; 62;
3: 147; 143; 142; 
10: 521; 444; 441;

galaxy s22 (high-end mobile), chrome 120:
1: 63; 71; 62;
3: 153; 147; 141; 
10: 519; 482; 480;

pixel 6, chrome 120:
1: 46; 27; 25;
3: 77; 70; 69;
10: 226; 239; 227;

iphone 13, safari 17:
1: 23; 19; 18;
3: 28; 32; 30;
10: 68; 67; 66;

iphone 13, chrome 120:
1: 19; 13; 14;
3: 30; 27; 28;
10: 66; 67; 68;



Overall, iphone seems to have no performance issues regarding DOMParser, although samsung devices are really slow in it; Since DOMParser is synchronous and it blocks further execution of script, it can have significant performance impact on certain devices

**Comment by hirasso** [comment]
Thanks for the performance test results @ScoobyDid ! 

I'd think a `visit.to.document` that is being cleaned up after `visit:end` would be nice. I can also imagine that with the recent improvements for [rapid navigation](https://github.com/swup/swup/releases/tag/4.5.0), this could become a little challenging. 

Let's find out!

**Comment by ScoobyDid** [comment]
@hirasso Thanks for the offer! Of course i may try to, but since i am not really into the context of swup (i came across it just a week ago or so and have reviewed only a tiny part of the source code), if i manage to do it, i may introduce many bugs that would be pain to fix.

For now, i can do a simpler PR regarding possible bug in parallel plugin - sometimes styles are not applied to .is-next-container; it works correctly in [your demo](https://replit.com/@swupjs/Swup-Demo-Reveal#base.css) on replit, but have inconsistent behavior in [same demo](https://stackblitz.com/edit/web-platform-bmtpwk?file=index.html,page-2.html,page-3.html,page-4.html,base.css,pages.css,transitions.css,swup.js) on stackblitz, and in my local environment too; i suppose the problem is in forceReflow helper; i will investigate it further tomorrow, and if it is indeed a bug, i will try to fix it.

**Comment by daun** [comment]
@ScoobyDid Curious what issues you're seeing with Parallel Plugin, but it's probably best to open a separate issue on the plugin repo so we can keep track of it there.

**Comment by daun** [comment]
@ScoobyDid Thanks for the suggestion! We've implemented this in #869.

**Comment by ScoobyDid** [comment]
Awesome! Thank you guys for such a superb library ✨

**Comment by hirasso** [comment]
Released in [version 4.6.0](https://github.com/swup/swup/releases/tag/4.6.0)
